geps/gep-1867.md

```diff
@@ -0,0 +1,128 @@
+
+### Dynamic Changes
+
+Currently, the spec recommends `GatewayClass` to be used as a *template*.
+Changes to it are not expected to change deployed `Gateway`s.
+
+This makes usage problematic in a declarative way.
+For example, if I wanted to represent a `version` field and change that to trigger an upgrade, I would need to create an entirely new `Gateway`.
+
+## API
+
+In order to address the concerns above, I propose a standard `infrastructure` API is added to `Gateway` and `GatewayClass`.
+Note the important part of this is the `Gateway` change; the `GatewayClass` aspect is mostly for consistency.
+
+Note that many of the fields under `GatewayInfrastructure` refer to work-in-progress GEPs.
+These fields are not tied to this GEP, and will only be included if the respective GEPs are approved.
+However, they are shown below to give a sense of the use cases the field can support.
+
+```go
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+
+type GatewayClassSpec struct {
+  // Infrastructure defines infrastructure level attributes for all Gateways in this class.
+  // A Gateway may provide configuration for the same values; as all fields in GatewayInfrastructure are implementation specific,
+  // the merging logic between these is as well. However, the GatewayClass is generally expected to be providing defaults
+  // rather than overrides
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+
+type GatewayInfrastructure struct {
+  // AttachTo marks this Gateway as a child to the referenced parent Gateway.
+  // See GEP-1713 for details, which this depends on.
+  AttachTo LocalObjectReference
+
+  // Scope defines the "scope" of the Gateway's addresses.
+  // Valid options are ClusterLocal or Public
+  // See GEP-1651 for details, which this depends on.
+  // NOTE: that GEP is likely to be implemented in a different way, at which point this field can be removed from this struct.
+  Scope AddressScope
+
+  // Version defines the version of the Gateway infrastructure to use.
+  // This is an opaque string that is implementation specific.
+  // Some possible valid examples include "canary", "v1.2.3", "docker.io/example:latest".
+  Version string
+
+  // Size defines the size of the Gateway infrastructure.
+  // The meaning of this is implementation specific.
+  // GEP-1762 will recommend this to mean "Pod Replicas" for in-cluster deployments.
+  Size int
+
+  // ParametersRef provides a arbitrary implementation-specific configuration for
+  // fields not expressed directly in this struct.
+  // This follows the same semantics as GatewayClass's ParametersRef, but lives on the Gateway.
+  ParametersRef ParametersReference
```

Comment:
I'm having trouble determining if some of these fields are covered by separate GEPs or meant to be included here. One quick comment - if infrastructure is included on GatewayClass, we could not include ParametersRef in the struct as it would conflict with the existing field.

Reply:
Good point. We can probably do a GatewayInfrastructure and GatewayClassInfrastructure, with GatewayInfrastructure just having additional ParametersRef that is not in GatewayClassInfrastructure to address this.

or we deprecate GatewayClass.ParametersRef, but that is probably not worth it

Reply:
We don't want AttachTo in GatewayClassInfrastructure either

Reply:
I removed all the fields to avoid confusion